amaranth/vendor/lattice_nexus.py

```diff
@@ -0,0 +1,748 @@
+                    raise ValueError(
+                        "OSCA divider (osca_div) must be an integer between 2 "
+                        "and 256"
+                    )
+                if (
+                    not isinstance(self.osca_div, int)
+                    or self.osca_div < 2
+                    or self.osca_div > 256
+                ):
+                    raise ValueError(
+                        "OSCA divider (osca_div) must be an integer between 2 "
+                        "and 256, not {!r}".format(self.osca_div)
+                    )
+                clk_i = Signal()
+                m.submodules += Instance(
+                    "OSCA",
+                    p_HF_CLK_DIV=str(self.osca_div - 1),
+                    i_HFOUTEN=Const(1),
+                    i_HFSDSCEN=Const(0),  # HFSDSCEN used for SED/SEC detector
+                    o_HFCLKOUT=clk_i,
+                )
+            else:
+                clk_i = self.request(self.default_clk).i
+            if self.default_rst is not None:
+                rst_i = self.request(self.default_rst).i
+            else:
+                rst_i = Const(0)
+
+            gsr0 = Signal()
+            gsr1 = Signal()
+            # On Nexus all the D-type FFs have either an synchronous or asynchronous preset. Here
+            # we build a simple reset synchronizer from D-type FFs with a positive-level
+            # asynchronous preset which we tie low
+            m.submodules += [
+                Instance(
+                    "FD1P3BX",
+                    p_GSR="DISABLED",
+                    i_CK=clk_i,
+                    i_D=~rst_i,
+                    i_SP=Const(1),
+                    i_PD=Const(0),
+                    o_Q=gsr0,
+                ),
+                Instance(
+                    "FD1P3BX",
+                    p_GSR="DISABLED",
+                    i_CK=clk_i,
+                    i_D=gsr0,
+                    i_SP=Const(1),
+                    i_PD=Const(0),
+                    o_Q=gsr1,
+                ),
+                Instance("GSR", p_SYNCMODE="SYNC", i_CLK=clk_i, i_GSR_N=gsr1),
+            ]
+            # GSR implicitly connects to every appropriate storage element. As such, the sync
+            # domain is reset-less; domains driven by other clocks would need to have dedicated
+            # reset circuitry or otherwise meet setup/hold constraints on their own.
+            m.domains += ClockDomain("sync", reset_less=True)
+            m.d.comb += ClockSignal("sync").eq(clk_i)
+            return m
+
+    # pg. 59 family datasheet
+    _single_ended_io_types = [
+        "LVCMOS33", "LVTTL33", 
+        "LVCMOS25", 
+        "LVCMOS18", "LVCMOS18H",
+        "LVCMOS15", "LVCMOS15H",  
+        "LVCMOS12", "LVCMOS12H", 
+        "LVCMOS10", "LVCMOS10H", "LVCMOS10R", 
+        "SSTL135_I", "SSTL135_II", "SSTL15_I", "SSTL15_II",
+        "HSUL12", "MIPI D-PHY LP Input (try using LVCMOS12)"
```

Comment:
> "MIPI D-PHY LP Input (try using LVCMOS12)"

Is this an actual valid value?

Reply:
I doubt it and in any case this IO_TYPE isn't mentioned in the datasheet page 59 (FPGA-DS-02078-1.6). @slagernate do you know where this came from?

Grep'ing through the IP blocks bundled with radiant it seems that `MIPI_DPHY` (no dash) could be a value IO_TYPE, but that would belong in the `_differential_io_types` list below.

Reply:
My apologies. ([see my comment here for a bit more context](https://github.com/amaranth-lang/amaranth/pull/758#discussion_r1123659819)). 

Pg. 59 was a reference to the CrossLink-NX datasheet. I see you referenced the Certus-NX datasheet. Probably it makes sense if we use the Nexus SysIO (FPGA-TN-02067-1-8-sysIO-User-Guide-Nexus-Platform.pdf). 

I will update this to remove `MIPI D-PHY LP Input (try using LVCMOS12)` and to conform to [Table 6.1 in the Nexus SysIO datasheet](https://pasteboard.co/uFsKvtlNhu1D.png), which should comply with the [nexus nextpnr database](https://github.com/YosysHQ/nextpnr/blob/2f509734dff676b7543a574b64a46be4224a5aa4/nexus/io.cc#L26) as well. (I intended to do this before but it slipped through the cracks).

Reply:
Sorry my messages were delayed (I guess I needed to start review). Happy to do more testing or take feedback to see this merged. Thanks